common/src/main/java/net/neoforged/gradle/common/extensions/JavaVersionManager.java

```diff
@@ -0,0 +1,29 @@
+package net.neoforged.gradle.common.extensions;
+
+import org.gradle.api.Project;
+import org.gradle.api.plugins.JavaPluginExtension;
+import org.gradle.jvm.toolchain.JavaLanguageVersion;
+import org.gradle.jvm.toolchain.internal.JavaToolchain;
+
+import javax.inject.Inject;
+
+public class JavaVersionManager {
+
+    private final Project project;
+    private int javaVersion = -1;
+
+    @Inject
+    public JavaVersionManager(Project project) {
+        this.project = project;
+    }
+
+    public void setJavaVersion(int javaVersion, String context) {
+        if (this.javaVersion > javaVersion) {
```

Comment:
The logic needs to be the other way around. In the case where there's 2 neo deps in the same project, the project must compile with the lowest Java version so that both sourcesets will work. Compiling with the highest will cause the one needing the lower to crash at runtime, and even if target compatibility were set, one can easily run into the issue of using newer Java methods.

Reply:
The issue is that the toolchain is used for both running. And NG doesn't currently have an easy way to use a separate javalauncher for a given run configuration it seems? Maybe the real solution would be to add that.

Reply:
Just to clarfy, not just running, also decompile and recompile need the newest versions, so this is the correct default behaviour.